### PR TITLE
fixed masked padding bug.

### PR DESCRIPTION
**Pull request by A-Jacobson**
when `used_masked_padding` is `False`, make sure the attention mask passed the `unet` at generate time is `None`. 

Also added this flag to our test suite so we can catch problems with it in the future.

### CODE REVIEW

File: diffusion/models/stable_diffusion.py [hunk 1]
```diff
@@ -469,22 +462,16 @@ def _prepare_text_embeddings(self, prompt, tokenized_prompts, tokenized_pad_mask
...
-                for i in range(1, tokenized_pad_mask.shape[1]):
-                    attention_mask |= tokenized_pad_mask[:, i]
-                tokenized_pad_mask = attention_mask
-            tokenized_pad_mask = tokenized_pad_mask.repeat(1, num_images_per_prompt, 1)
-            tokenized_pad_mask = tokenized_pad_mask.view(bs_embed * num_images_per_prompt, seq_len)  # [B, 77]
+        prompt_embeds = _duplicate_tensor(prompt_embeds, num_images_per_prompt)
+
+        if self.mask_pad_tokens:
+            tokenized_pad_mask = _create_unet_attention_mask(tokenized_pad_mask)
+            tokenized_pad_mask = _duplicate_tensor(tokenized_pad_mask, num_images_per_prompt)
```
Comment (Landanjs) [inline diffusion/models/stable_diffusion.py:469]: I just realized that my code is not be right 😬 When duplicating with num_images_per_prompt, the duplicated prompt embeds will be consecutive, but this is not true for the tokenized_pad_mask due to the different dimensions. 

Two fixes come to mind:
1. quickest fix would be to `tokenized_pad_mask.unsqueeze(-1)`
2. The nasty looking but feels more principled fix in `_duplicate_tensor`: `tensor = tensor.repeat(1, num_images_per_prompt, *[1,]*len(tensor.shape[2:]))` and `tensor.view(batch_size * num_images_per_prompt, seq_len, *[-1,]*len(tensor.shape[2:]))`
Using the second means you can avoid the `.squeeze(-1)`

Results:
```
>>> t2 = torch.randn(2, 3, 4)
>>> t2
tensor([[[-1.4956,  0.5156,  1.3679, -0.6651],
         [ 0.5277,  0.8577,  3.6882,  1.7380],
         [ 0.7071, -1.2043,  0.6128, -0.3668]],

        [[ 0.3232,  0.3563,  0.0479,  0.2255],
         [-0.5160, -0.2547,  0.8563,  1.2211],
         [-0.3290,  0.5519,  0.0721, -0.2651]]])
>>> t2.repeat(1, 2, 1).view(2*2, 3, -1)
tensor([[[-1.4956,  0.5156,  1.3679, -0.6651],
         [ 0.5277,  0.8577,  3.6882,  1.7380],
         [ 0.7071, -1.2043,  0.6128, -0.3668]],

        [[-1.4956,  0.5156,  1.3679, -0.6651],
         [ 0.5277,  0.8577,  3.6882,  1.7380],
         [ 0.7071, -1.2043,  0.6128, -0.3668]],

        [[ 0.3232,  0.3563,  0.0479,  0.2255],
         [-0.5160, -0.2547,  0.8563,  1.2211],
         [-0.3290,  0.5519,  0.0721, -0.2651]],

        [[ 0.3232,  0.3563,  0.0479,  0.2255],
         [-0.5160, -0.2547,  0.8563,  1.2211],
         [-0.3290,  0.5519,  0.0721, -0.2651]]])
>>> t3 = torch.randn(2, 3)
>>> t3
tensor([[-0.5588, -0.9764, -1.1396],
        [ 1.1794,  0.9496, -0.2843]])
>>> t3.repeat(1, 2, 1).view(4, 3, -1).squeeze(-1)
tensor([[-0.5588, -0.9764, -1.1396],
        [ 1.1794,  0.9496, -0.2843],
        [-0.5588, -0.9764, -1.1396],
        [ 1.1794,  0.9496, -0.2843]])
```

File: diffusion/models/stable_diffusion.py [hunk 2]
```diff
@@ -469,22 +462,16 @@ def _prepare_text_embeddings(self, prompt, tokenized_prompts, tokenized_pad_mask
...
+        else:
+            tokenized_pad_mask = None  # don't use attention mask in unet
 
         if self.sdxl and pooled_text_embeddings is not None:
-            pooled_text_embeddings = pooled_text_embeddings.repeat(1, num_images_per_prompt)
-            pooled_text_embeddings = pooled_text_embeddings.view(bs_embed * num_images_per_prompt, -1)
+            pooled_text_embeddings = _duplicate_tensor(pooled_text_embeddings, num_images_per_prompt)
```
Comment (Landanjs) [inline diffusion/models/stable_diffusion.py:474]: Same issue as above, I think. You could also use a squeeze here or if you use the 2 fix, it should be fine

File: tests/test_model.py [hunk 3]
```diff
@@ -47,10 +47,15 @@ def test_sd2_generate(guidance_scale, negative_prompt):
     assert output.shape == (1, 3, 8, 8)
 
 
-def test_sdxl_forward():
+@pytest.mark.parametrize('mask_pad_tokens', [True, False])
```
Comment (Landanjs) [inline tests/test_model.py:50]: Should we also test this with the SD2 forward and generate?

**Review by Landanjs — COMMENTED** [review]
Thanks for fixing my mistake!! It looks like I made another mistake in the code... curious to hear which fix you think is better. If you'd prefer not to fix it, I can either adjust this PR or make a new PR

File: tests/test_model.py (hunk 3 above)
Comment (A-Jacobson) [reply tests/test_model.py:50]: I need to redo the tests so we don't build so many models first. A few more and the tests will start timing out again haha

Comment (Landanjs) [reply tests/test_model.py:50]: SG!

File: diffusion/models/stable_diffusion.py [hunk 4]
```diff
@@ -469,22 +462,16 @@ def _prepare_text_embeddings(self, prompt, tokenized_prompts, tokenized_pad_mask
...
-            if len(tokenized_pad_mask.shape) == 3:
-                attention_mask = tokenized_pad_mask[:, 0]
-                for i in range(1, tokenized_pad_mask.shape[1]):
-                    attention_mask |= tokenized_pad_mask[:, i]
-                tokenized_pad_mask = attention_mask
-            tokenized_pad_mask = tokenized_pad_mask.repeat(1, num_images_per_prompt, 1)
-            tokenized_pad_mask = tokenized_pad_mask.view(bs_embed * num_images_per_prompt, seq_len)  # [B, 77]
+        prompt_embeds = _duplicate_tensor(prompt_embeds, num_images_per_prompt)
+
+        if self.mask_pad_tokens:
```
[the remark this answers is not in the document]
Comment (A-Jacobson) [reply diffusion/models/stable_diffusion.py:467]: I guess? I can add another test for unconditional generation to check for this.

[the remark this answers is not in the document]
Comment (A-Jacobson) [reply diffusion/models/stable_diffusion.py]: misunderstood the bug, it's not with unconditional generation. We now check that `tokenized_pad_mask` exists before trying to do stuff to it. Nice catch.

File: diffusion/models/stable_diffusion.py (hunk 1 above)
Comment (A-Jacobson) [reply diffusion/models/stable_diffusion.py:472]: flipped a count and went with #2 =)

**Review by Landanjs — APPROVED** [review]
LGTM!!
